File: game-of-life-cpp/Dandelion.cpp
#include "include/Dandelion.h"

Dandelion::Dandelion(World* world, Point& position)
    : Plant(world, 0, position, "🌽", 0) {
}

Dandelion::Dandelion(World* world, Point& position, int age)
    : Plant(world, 0, position, "🌽", age) {
}

Dandelion::~Dandelion() {

}

void Dandelion::action() {
    int i = 0;
    while(i < 3) {
        if (canReproduceThisTurn() && hasFreeSpace() && age != 0) {
            reproduce(this->position);
            break;
        }
        i++;
    }
    age++;
}

bool Dandelion::canReproduceThisTurn() const {
    return (rand() % 32 == 0);
}

File: game-of-life-java/src/Dandelion.java
import javax.swing.*;
import java.awt.*;

public class Dandelion extends Plant {
    private static final ImageIcon dandelionIcon = new ImageIcon("images/dandelion.png");
    public static final Image scaledDandelionIcon = dandelionIcon.getImage().getScaledInstance(30, 30, Image.SCALE_SMOOTH);

    public Dandelion(Point position, World world, int age) {
        super(0, position, world, age);
    }

    @Override
    public ImageIcon getIcon() {
        return dandelionIcon;
    }

    @Override
    protected boolean canReproduceThisTurn() {
        return (Math.random() < 0.03);
    }

    @Override
    public void action() {
        int i = 0;
        while(i < 3) {
            if(canReproduceThisTurn() && hasFreeSpace() && age != 0) {
                reproduce(position);
                break;
            }
            i++;
        }
        age++;
    }
}
